scipy/stats/_qmc.py

```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+    >>> sampler = qmc.OptimalLatinHypercube(d=2)
+    >>> sample = sampler.random(n=5)
+    >>> sample
+    array([[0.84052691, 0.53664833],  # random
+           [0.1545328 , 0.36265316],
+           [0.52177809, 0.93343721],
+           [0.68033825, 0.06474907],
+           [0.26544879, 0.61163943]])
+
+    Compute the quality of the sample using the discrepancy criterion.
+
+    >>> qmc.discrepancy(sample)
+    0.01602810655510689  # random
+
+    You can possibly improve the quality of the sample by performing more
+    optimization iterations by using `niter`:
+
+    >>> sampler_2 = qmc.OptimalLatinHypercube(d=2, n_perturbations=5000)
+    >>> sample_2 = sampler_2.random(n=5)
+    >>> qmc.discrepancy(sample_2)
+    0.01600398742881648  # random
+
+    """
+
+    def __init__(self, d: IntNumber, *, start_sample: npt.ArrayLike = None,
+                 n_perturbations: IntNumber = 10000,
+                 seed: SeedType = None):
+        super().__init__(d=d, seed=seed)
+        self.n_perturbations = n_perturbations
+        self.start_sample: Optional[np.ndarray] = None
+
+        if start_sample is not None:
+            self.start_sample = np.asarray(start_sample).copy()
+            self.start_disc = discrepancy(self.start_sample)
+        else:
+            self.start_disc = np.inf
+
+        self.lhs = LatinHypercube(self.d, seed=self.rng)
+
+    def random(self, n: IntNumber = 1):
+        """Draw `n` in the half-open interval ``[0, 1)``.
+
+        Parameters
+        ----------
+        n : int, optional
+            Number of samples to generate in the parameter space. Default is 1.
+
+        Returns
+        -------
+        sample : array_like (n, d)
+            Optimal LHS sample.
+
+        """
+        if self.d == 0 or n == 0:
+            return np.empty((n, self.d))
+
+        if self.start_sample is None:
+            best_sample = self.lhs.random(n)
+            best_disc = discrepancy(best_sample)
+        else:
+            best_sample = self.start_sample
+            best_disc = self.start_disc
+
+        if n == 1:
+            return best_sample
+
+        bounds = ([0, self.d - 1],
+                  [0, n - 1],
+                  [0, n - 1])
+
+        for _ in range(self.n_perturbations):
+            col = rng_integers(self.rng, *bounds[0])
+            row_1 = rng_integers(self.rng, *bounds[1])
+            row_2 = rng_integers(self.rng, *bounds[2])
+            doe = best_sample.copy()
+            disc = _perturb_discrepancy(best_sample,
+                                        row_1, row_2, col,
+                                        best_disc)
+            if disc < best_disc:
+                doe[row_1, col], doe[row_2, col] = (
+                    doe[row_2, col], doe[row_1, col])
+                best_disc = disc
+                best_sample = doe
+
+        self.num_generated += n
+        return best_sample
+
+    def reset(self):
+        """Reset the engine to base state.
```

Comment:
And if `start_sample` is provided, then `reset` does a shallow reset of the object to the initial state - shallow because it does not reset the underlying RNG unless an `int` is provided for `seed`. Correct? 

I'm thinking about how it can be tested. I guess don't see a real use case of the method unless `seed` is an `int`, so the test would just make sure that `reset` and a new object (constructed with a deep-copied RNG) do the same things.

Reply:
I removed this reset as per the other comment. It's not needed :)

Reply:
It is still inheriting the `reset` method from `QMCEngine`, but at least that creates a deep copy.